### ISSUE TITLE
Tally Arithmetic

### ISSUE THREAD

**Issue by wbinventor** [mit-crpg/openmc#383]
For tally data analysis, it would be helpful to extend the Python API with a "tally arithmetic" scheme akin to what I implemented for [PINSPEC](https://github.com/mit-crpg/PINSPEC). The idea behind tally arithmetic is that it allows users to perform numerical operations (+ - / *) directly between NumPy arrays and/or scalar-valued data. This would require that the ``Tally`` class in the Python API be extended to include new ``__add__(...)``, ``__sub__(...)``, ``__mul__(...)`` and ``__div__(...)`` methods. Each of these methods would do the following

* analyze the tally data (or scalar value) for the LHS and RHS operands
* apply the arithmetic operation between the tally means to compute a derived tally metric
* use [error propagation theory](http://en.wikipedia.org/wiki/Taylor_expansions_for_the_moments_of_functions_of_random_variables) to estimate the variance for the derived metric

The final step would be the most complicated (but useful) component to the tally arithmetic concept. Of course some assumption(s) about the covariance matrix between the LHS and RHS tally bins would have to be made for this to work. For example, the naive assumption that tally bins are independent random variables would allow the off-diagonal terms to be set to zero for the most simple calculation.

Below is an example Python snippet for how I envision that tally arithmetic might work to compute derived tally metrics like multi-group cross-sections:

```python
from openmc.statepoint import StatePoint

# Build StatePoint file and extract tally data
sp = StatePoint('statepoint.XXX.h5')
sp.read_results()

# Get Tally objects for flux and absorption rate
mesh_id = sp._meshes.keys()[0]
mesh = openmc.Filter(type='mesh', bins=[mesh_id])
energy = openmc.Filter(type='energy', bins=[0., 0.625e-6, 20.])
flux = sp.get_tally(score='flux', filters=[mesh, energy], name='flux')
flux = sp.get_tally(score='flux', filters=[mesh, energy], name='abs rate')

# TALLY ARITHMETIC
# Compute multi-group cross-sections into a "DerivedTally" object
abs_xs = abs_rate / flux

# Do more cool stuff
...
```

This extension would likely builid off of #382 for tallies with multiple scores. As with #382, this feature would greatly simplify my personal data workflow, but given my schedule righ now, I'm just throwing this idea up here in case anyone has any feedback before I set about implementing it.

**Comment by walshjon** [comment]
Bullet 3 does look to be the sticking point.  PINSPEC currently [uses](https://github.com/mit-crpg/PINSPEC/commit/4cc730543d7925f89f02764f25c297bffcc34f69) a zeroth order Taylor expansion for error propagation.  This seems to be as good as we can do without correlation coefficients between the tallies of interest.  As long as the assumption of tally independence is spelled out upfront, though, this seems fine.

**Comment by nelsonag** [comment]
I've been using the uncertainties package to do the third bullet for me (
https://pythonhosted.org/uncertainties/).

**Comment by wbinventor** [comment]
Yeah I used the ``uncertainties`` package for some of my downstream data
processing as well. That is until I discovered that its error propagation
calculation for the variance is insanely slow. What's more, it scales as
something like O(N^3) with N tally bins from my rough empirical analysis.
This can he particularly problematic for large tallies with  "mesh" or
"distribcell" filters.

I raised this issue with the core developer but it's not a priority for him
as the code is for all practical purposes dead now. Fortunately, it is
straightforward to directly implement the zeroth order error propagation
calculation with vectorizes NumPy arithmetic. I did this and it performs
much better than ``uncertainties`` and can easily be transferred into the
tally arithmetic scheme. Of course the downside is that it does not
leverage the nice pretty printing utilities available with
``uncertainties``, but if Pandas is used per #382 then this will be a mute
point.
